ambry-cloud/src/main/java/com.github.ambry.cloud/HelixVcrCluster.java

```diff
@@ -0,0 +1,130 @@
+  private final HelixVcrClusterMetrics metrics;
+
+  /**
+   * Construct the helix VCR cluster.
+   * @param cloudConfig The cloud configuration to use.
+   * @param clusterMapConfig The clustermap configuration to use.
+   * @param clusterMap The clustermap to use.
+   */
+  public HelixVcrCluster(CloudConfig cloudConfig, ClusterMapConfig clusterMapConfig, ClusterMap clusterMap)
+      throws Exception {
+    if (Utils.isNullOrEmpty(cloudConfig.VCR_CLUSTER_ZK_CONNECT_STRING)) {
+      throw new IllegalArgumentException("Missing value for " + CloudConfig.VCR_CLUSTER_ZK_CONNECT_STRING);
+    } else if (Utils.isNullOrEmpty(cloudConfig.VCR_CLUSTER_NAME)) {
+      throw new IllegalArgumentException("Missing value for " + CloudConfig.VCR_CLUSTER_NAME);
+    } else if (Utils.isNullOrEmpty(cloudConfig.VCR_CLUSTER_ZK_CONNECT_STRING)) {
+      throw new IllegalArgumentException("Missing value for " + CloudConfig.VCR_CLUSTER_ZK_CONNECT_STRING);
+    }
+
+    currentDataNode = new CloudDataNode(cloudConfig, clusterMapConfig);
+    List<? extends PartitionId> allPartitions = clusterMap.getAllPartitionIds(null);
+    logger.trace("All partitions from clusterMap: " + allPartitions);
+    partitionIdMap = allPartitions.stream().collect(Collectors.toMap(PartitionId::toPathString, Function.identity()));
+    vcrClusterName = cloudConfig.vcrClusterName;
+    vcrInstanceName =
+        ClusterMapUtils.getInstanceName(clusterMapConfig.clusterMapHostName, clusterMapConfig.clusterMapPort);
+
+    manager = HelixManagerFactory.getZKHelixManager(vcrClusterName, vcrInstanceName, InstanceType.PARTICIPANT,
+        cloudConfig.vcrClusterZkConnectString);
+    manager.getStateMachineEngine()
+        .registerStateModelFactory(LeaderStandbySMD.name, new HelixVcrStateModelFactory(this));
+    manager.connect();
+    helixAdmin = manager.getClusterManagmentTool();
+    metrics = new HelixVcrClusterMetrics(clusterMap.getMetricRegistry());
+    logger.info("HelixVcrCluster started successfully.");
+  }
+
+  synchronized public void addPartition(String partitionIdStr) {
+    if (partitionIdMap.containsKey(partitionIdStr)) {
+      assignedPartitionIds.add(partitionIdMap.get(partitionIdStr));
+      logger.info("Added partition {} to current VCR: ", partitionIdStr);
+    } else {
+      logger.trace("Partition {} not in clusterMap on add.", partitionIdStr);
+      metrics.partitionIdNotInClusterMapOnAdd.inc();
+    }
+  }
+
+  synchronized public void removePartition(String partitionIdStr) {
+    if (partitionIdMap.containsKey(partitionIdStr)) {
+      assignedPartitionIds.remove(partitionIdMap.get(partitionIdStr));
+      logger.info("Removed partition {} to current VCR: ", partitionIdStr);
+    } else {
+      logger.trace("Partition {} not in clusterMap on remove.", partitionIdStr);
+      metrics.partitionIdNotInClusterMapOnRemove.inc();
+    }
+  }
+
+  @Override
+  public List<? extends DataNodeId> getAllDataNodeIds() {
+    // TODO: return all VCR nodes for recovery.
```

Comment:
Can we get this list from the Helix manager currently?  If not, what is needed for that?

Reply:
It takes some time, but it's used for recovery(for server to get a list of vcr host). I can do the code change as follow up PR.

Reply:
ok